addons/mass_mailing/models/mailing_list.py

```diff
@@ -136,3 +192,109 @@ def action_merge(self, src_lists, archive):
 
     def close_dialog(self):
         return {'type': 'ir.actions.act_window_close'}
+
+    # ------------------------------------------------------
+    # UTILITY
+    # ------------------------------------------------------
+
+    def _fetch_contact_nbr_counts(self):
+        """ Compute number of contacts matching various conditions.
+        (see '_get_contact_nbr_select_fields' for details)
+
+        Will return a dict under the form:
+        {
+            42: { # 42 being the mailing list ID
+                'contact_nbr': 52,
+                'contact_nbr_email': 35,
+                'contact_nbr_opt_out': 5,
+                'contact_nbr_blacklisted': 2
+            },
+            ...
+        } """
+
+        res = []
+        if self.ids:
+            self.env.cr.execute(f'''
```

Comment:
Could probably beneficts from some reorganize, like
1/ compute select information (with list_id in the sub method)
2/ compute from information - maybe only join information (with "left join mailing_contact" in sub method)
3/ write SQL query with select and from using join

That way I think it would be clearer when reading code: all/most information are present in sub methods and main method does only some formatting. Currently we have a bit of both and it makes understanding code really hard.